src/components/views/rooms/VoiceRecordComposerTile.tsx

```diff
@@ -39,25 +46,67 @@ interface IProps {
 interface IState {
     recorder?: VoiceRecording;
     recordingPhase?: RecordingState;
+    relHeights: number[];
+    seconds: number;
 }
 
 /**
  * Container tile for rendering the voice message recorder in the composer.
  */
 @replaceableComponent("views.rooms.VoiceRecordComposerTile")
 export default class VoiceRecordComposerTile extends React.PureComponent<IProps, IState> {
+    private waveform: number[] = [];
+    private seconds = 0;
+    private scheduledAnimationFrame = false;
+
     public constructor(props) {
         super(props);
 
         this.state = {
             recorder: null, // no recording started by default
+            seconds: 0,
+            relHeights: arraySeed(0, RECORDING_PLAYBACK_SAMPLES),
         };
     }
 
+    public componentDidUpdate(prevProps, prevState) {
+        if (!prevState.recorder && this.state.recorder) {
+            this.state.recorder.liveData.onUpdate(this.onRecordingUpdate);
+        }
+    }
+
     public async componentWillUnmount() {
         await VoiceRecordingStore.instance.disposeRecording();
     }
 
+    private onRecordingUpdate = (update: IRecordingUpdate): void => {
+        this.waveform = update.waveform;
+        this.seconds = update.timeSeconds;
+
+        if (this.scheduledAnimationFrame) {
+            return;
+        }
+
+        this.scheduledAnimationFrame = true;
+        // The audio recorder flushes data faster than the screen refresh rate
+        // Using requestAnimationFrame makes sure that we only flush the data
+        // to react once per tick to avoid unneeded work.
+        requestAnimationFrame(() => {
+            // The waveform and the downsample target are pretty close, so we should be fine to
+            // do this, despite the docs on arrayFastResample.
+            const bars = arrayFastResample(Array.from(this.waveform), RECORDING_PLAYBACK_SAMPLES);
+            this.setState({
+                // The incoming data is between zero and one, but typically even screaming into a
+                // microphone won't send you over 0.6, so we artificially adjust the gain for the
+                // waveform. This results in a slightly more cinematic/animated waveform for the
+                // user.
+                relHeights: bars.map(b => percentageOf(b, 0, 0.50)),
+                seconds: this.seconds,
+            });
+            this.scheduledAnimationFrame = false;
+        });
+    }
```

Comment:
This all needs to move into the recording waveform too, as part of that component structure. I'm not seeing a reason why the clock needs to be counted as part of this, so should be fine to leave that one with the regular de-duped state updates?

Reply:
To make this a bit more reusable in other components, we can leverage the `MarkedExecution` util we have, I think.

It needs a small update in order to work though: https://github.com/matrix-org/matrix-react-sdk/commit/824d92d22cecb1148a43b645ae61d83d57176250

Usage looks something like this: https://github.com/matrix-org/matrix-react-sdk/blob/ff02a41cfba89fd6454c8ec11a6e27a5b5f471d7/src/components/views/audio_messages/SeekBar.tsx#L50-L76

(these bits are part of an upcoming PR, but 824d92d22cecb1148a43b645ae61d83d57176250 should be cherrypickable)